### ISSUE TITLE
Script injection not working if head tag is missing

### ISSUE THREAD

**Issue by gkutiel** [cortesi/devd#65]
If the served html file does not have a head tag then the live reload script is not injected. I think it is enough to document this.

This is a great tool, thank you.

**Comment by cortesi** [comment]
Thanks for the feedback. I've added a note clarifying this to the README.

**Comment by cornop** [comment]
@cortesi Is this not something that would be fixable by devd? It's an odd gotcha, especially considering that [the Google Style Guide recommends omitting the header tag](https://stackoverflow.com/a/25749523).

**Comment by cortesi** [comment]
Yes, I'd certainly accept a patch that adds the header tag if it's missing.

**Comment by cortesi** [comment]
Adding a bit more nuance here:

- I don't think it's common practice to omit the head tag, and I don't think that's what the style guidance you linked to recommends. That said, it's common enough that we should cater for it.
- I think we can amend the behaviour to try to inject into the head tag, and if that can't be done, to inject our own head tag just after the opening "html" tag. If the html tag is not found either, we fail. All this has to be bounded, and 30k is a good upper limit.
- Doing this right would require some careful work in the injection mechanism. Please have a chat with me if you intend to hack on it.

**Comment by cornop** [comment]
@cortesi I believe developers who leave out the head tag would also leave out the html tag. Injecting a head tag also sounds invasive.

Could we fallback to injecting the script
1. after the last meta tag, or
2. before the first script tag, or
3. after the doctype declaration?

**Comment by cortesi** [comment]
The html tag is not optional - if it's not there, then we're not dealing with HTML, and can't inject anything. I think there's a higher likelihood that injecting content into the body of the document would result in unanticipated behaviour.

**Comment by cornop** [comment]
The html tag *is* optional, and is also recommended to be omitted in the document I linked above. Ideally devd should follow the [HTML 5 rules for determining content types](https://dev.w3.org/html5/cts/html5-type-sniffing.html#content-type5), i.e. check for a doctype declaration.

Anyway, I'll drop this discussion now. I think I got my point across and I understand if you don't want to complicate your injection logic or change it in a backwards-incompatible way.
